boto/s3/checksums.py

```diff
@@ -127,25 +127,31 @@ def crc32c_cal(data):
   """
   return crc32c_finalize(crc32c_update(CRC32_INIT, data))
 
-def cal_checksum(fp, size=-1, type='crc32'):
+def cal_checksum(fp=None, size=-1, ctype='crc32', rdata=None):
     checksum = None
-    cpos = fp.tell()
-    rdata = fp.read(size)
-    if type == 'crc32':
+    if rdata is not None:
+        rdata = rdata
```

Comment:
What do these 2 lines do?

Reply:
Does letting rdata remain None not cause the checksum calculations to fail below if fp is also None?

Reply:
Thanks. Updated [here](https://github.com/cloudian/boto/pull/38/commits/334f657a89ad45ab928d195d4c30e21326682585)